### PR TITLE
Update multi() to work with non-std shapes

### PR DESCRIPTION
**Pull request by umangyadav**
Solves #1311

### CODE REVIEW

File: src/targets/cpu/include/migraphx/cpu/pointwise.hpp [hunk 1]
```diff
@@ -43,7 +43,8 @@ struct multi_index
     {
```
Comment (umangyadav) [inline src/targets/cpu/include/migraphx/cpu/pointwise.hpp:43]: This `multi_index` struct is not used anywhere for the CPU. 

Should we just remove that ?

Comment (pfultz2) [reply src/targets/cpu/include/migraphx/cpu/pointwise.hpp:43]: Its used below.

Comment (umangyadav) [reply src/targets/cpu/include/migraphx/cpu/pointwise.hpp:43]: I don't find any reference when I search for its constructor or when i search for "multi_index"

File: src/shape.cpp [hunk 2]
```diff
@@ -361,31 +361,25 @@ std::size_t shape::index(std::size_t i) const
     }
 }
 
-std::vector<std::size_t> shape::multi(std::size_t i) const
+std::vector<std::size_t> shape::multi(std::size_t idx) const
 {
-    assert(this->standard());
-
+    if(idx >= elements())
```
Comment (CharlieL7) [inline src/shape.cpp:366]: Add a throw for if the shape is dynamic.

**Review by CharlieL7 — CHANGES_REQUESTED** [review]

File: src/shape.cpp (hunk 2 above)
Comment (umangyadav) [reply src/shape.cpp:366]: Call to `elements()` would throw in that case. I can update test.

**Review by CharlieL7 — APPROVED** [review]

File: src/targets/cpu/include/migraphx/cpu/pointwise.hpp [hunk 3]
```diff
@@ -43,7 +43,8 @@ struct multi_index
...
         assert(n < max_size);
         std::copy(s.lens().begin(), s.lens().end(), dims);
-        s.multi_copy(i, index, index + max_size);
+        const std::vector<size_t> m_idx = s.multi(i);
+        std::copy(m_idx.begin(), m_idx.end(), index);
```
Comment (pfultz2) [inline src/targets/cpu/include/migraphx/cpu/pointwise.hpp:47]: This should use `multi_copy` so we dont need to construct a temporary vector(which uses heap allocation). The `index` member is an array so we dont need to do heap allocation when constructing.

File: src/shape.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (umangyadav) [reply src/shape.cpp:366]: Why is that ? Shouldn't it be illegal during runtime as well ?

File: src/targets/cpu/include/migraphx/cpu/pointwise.hpp (hunk 3 above)
Comment (umangyadav) [reply src/targets/cpu/include/migraphx/cpu/pointwise.hpp]: Added back.

File: src/shape.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pfultz2) [reply src/shape.cpp]: We shouldn't do this extra comparison when its not necessary. Calling with the correct index is a precondition of the function.

File: src/shape.cpp [hunk 4]
```diff
@@ -363,21 +363,23 @@ std::size_t shape::index(std::size_t i) const
 
 std::vector<std::size_t> shape::multi(std::size_t idx) const
 {
-    if(idx >= elements())
-    {
-        MIGRAPHX_THROW("Can not convert element-index: " + std::to_string(idx) +
-                       " to multi-dimensional index for shape with num elements: " +
-                       std::to_string(elements()));
-    }
+    assert(idx < elements());
```
Comment (pfultz2) [inline src/shape.cpp:366]: The assert should probably go into `multi_copy` as well.

Comment (umangyadav) [reply src/shape.cpp:366]: Done.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**